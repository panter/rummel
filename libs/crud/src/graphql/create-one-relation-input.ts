import { Type } from '@nestjs/common';
import { ArgsType, Field, InputType } from '@nestjs/graphql';
import { ConnectRelationInput } from './generic-types';
import { typesCache } from './types-cache';
import { upsertInput } from './upsert-input';
import { getTypeName } from './utils';

export const createOneRelationInput = <T>(
  classRef: Type<T>,
  options?: {
    parentRef: Type<any>;
    hideCreate?: boolean;
    parentProperty?: string;
  },
) => {
  const typeName = getTypeName(classRef);
  const withoutTypeName = options?.parentRef
    ? `Without${getTypeName(options?.parentRef)}`
    : '';
  const name = `${typeName}CreateNestedOne${withoutTypeName}Input`;

  if (typesCache[name]) {
    return typesCache[name];
  }

  @ArgsType()
  @InputType(name, { isAbstract: false })
  class RelationInputArgsType {
    @Field(() => ConnectRelationInput, { nullable: true })
    connect: any;

    create: any;
  }

  typesCache[name] = RelationInputArgsType;

  if (!options?.hideCreate) {
    // call createInput after adding to cache because of recursive call
    const WhereInputType = upsertInput(classRef, {
      ignoreType: options?.parentRef,
    });
    Field(() => WhereInputType, { nullable: true })(
      RelationInputArgsType.prototype,
      'create',
    );
  }

  return RelationInputArgsType;
};
