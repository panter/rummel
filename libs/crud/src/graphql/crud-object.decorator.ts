import { EntityManager } from '@mikro-orm/postgresql';
import { Type } from '@nestjs/common';
import 'reflect-metadata';
import { getTypeName } from './utils';

export type InputResolverOptions<Entity> = {
  em: EntityManager;
  currentUser: any;
  currentOrmData?: Entity;
  type: Type<Entity>;
};
export type InputResolver<Entity = any, GqlInput = any> = (
  gqlInput: GqlInput,
  options: InputResolverOptions<Entity>,
) => Promise<void>;

/**
 * The idea is to cache the options for each type.
 * Ideally we would like to use Reflect.getMetadata but it
 * I ran into problems so i resorted to to a classic map cache approach.
 *
 */

const cache = new Map<Type<unknown>, CrudObjectOptions>();
const addToCache = (target: any, options: CrudObjectOptions) => {
  cache.set(target, options);
};

// const CRUD_OBJECT_METADATA_NAME = 'crud:object';

export type CrudObjectProps<Entity = any, GqlInput = any> = {
  alias?: Type<unknown>;
  createOne?: [InputResolver<Entity, GqlInput>];
  updateOne?: [InputResolver<Entity, GqlInput>];
  // findManyOne?: (all the necessary args) => void );
  // ... all the cruds overwride
};
export type CrudObjectOptions<Entity = any, GqlInput = any> = CrudObjectProps<
  Entity,
  GqlInput
> & {
  targetName: string;
  target: Type<unknown>;
};

export function CrudObject<Entity = any, GqlInput = any>(
  options: CrudObjectProps<Entity, GqlInput> = {},
) {
  return function (target: any) {
    const meta: CrudObjectOptions<Entity, GqlInput> = {
      ...options,
      target,
      targetName: getTypeName(target),
    };
    addToCache(target, meta);
    // Reflect.defineMetadata(
    //   CRUD_OBJECT_METADATA_NAME,
    //   meta,
    // );
  };
}

export function getCrudObjectOptions<DTO>(
  target: Type<DTO>,
): CrudObjectOptions | undefined {
  return cache.get(target);
  // const options = getPrototypeChain(target).reduce((options, Cls) => {
  //   if (cache.get(Cls.prototype)) {
  //     return cache.get(Cls.prototype);
  //   }
  //   return options;
  // }, [] as any);

  // return options;
}
