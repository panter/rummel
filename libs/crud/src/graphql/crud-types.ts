import { Type } from '@nestjs/common';

export type CrudGqlType<T extends string> = T extends '__-tsissue-__'
  ? never
  : any;
export type ClassName<T> = T extends { new (...args: any[]): any }
  ? T extends { entityName: infer N }
    ? N
    : never
  : never;

export type CrudEntityType<T, N> = Type<T> & { entityName: N };
