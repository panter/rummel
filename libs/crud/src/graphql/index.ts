export * from './upsert-input';
export * from './resolvers/create-one.resolver';
export * from './resolvers/update-one.resolver';
export * from './resolvers/find-one.resolver';
export * from './resolvers/find-many.resolver';
export * from './resolvers/delete-one.resolver';
export * from './resolvers/object-relation-resolver';
export * from './create-one-entity-args';
export * from './find-many-entity-args';
export * from './gql-filter-to-mikro-orm';
export * from './gql-upsert-input-to-mikro-orm';
export * from './generic-types';
export * from './crud-field.decorator';
