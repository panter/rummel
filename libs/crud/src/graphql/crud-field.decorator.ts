import { EntityManager, ObjectQuery } from '@mikro-orm/postgresql';
import { Type } from '@nestjs/common';
import { ReturnTypeFunc } from '@nestjs/graphql';
import { TypeOptions } from '@nestjs/graphql/dist/interfaces/type-options.interface';
import { uniqBy } from 'lodash';
import 'reflect-metadata';
// import { UserEntity } from '../../modules/user/models/user.entity';
import { CrudInfo, getPrototypeChain } from './utils';
import { AuthenticatedUser } from './types';

/**
 * The idea is to cache the options for each type.
 * Ideally we would like to use Reflect.getMetadata but it
 * I ran into problems so i resorted to to a classic map cache approach.
 *
 */

const cache = new Map<Type<unknown>, { [key: string]: CrudFieldOptions }>();
const addToCache = (target: any, key: string, options: any) => {
  const cached = cache.get(target);
  if (cached) {
    cached[key] = options;
  } else {
    cache.set(target, { [key]: options });
  }
};

// const CRUD_FIELD_METADATA_NAME = 'crud:field';
export type WhereFieldResolver<Entity = any, GqlWhere = any> = (
  ormQuery: ObjectQuery<Entity>,
  gqlWhere: GqlWhere,
  options: {
    currentUser?: AuthenticatedUser | null;
    key: string | number | symbol;
  },
) => void;

export type InputFieldResolverOptions<Entity> = {
  em: EntityManager;
  currentUser?: AuthenticatedUser | null;
  currentOrmData?: Entity;
  rootOrmData?: Entity;
  type: Type<Entity>;
  crudInfo: CrudInfo;
  skipInputResolver?: boolean;
  key: string | number | symbol;
};
export type InputFieldResolver<Entity = any, GqlInput = any> = (
  ormData: Entity,
  gqlInput: GqlInput,
  options: InputFieldResolverOptions<Entity>,
) => Promise<void>;

export type CrudRelationFieldOptions = {
  /**
   * If set to `true`, the relation cannot be connected by primary key
   */
  hideConnect?: boolean;
  /**
   *  If set to `true`, the relation can be created with all it's fields.
   */
  showCreate?: boolean;
  /**
   *  If set to `true`, the relation can be updated with all it's fields.
   */
  showUpdate?: boolean;
  /**
   *  If set to `true`, the relation cannot be disconnected by primary key.
   */
  hideDisconnect?: boolean;
};

export type CrudFieldProps<Entity = any, GqlInput = any, GqlWhere = any> = {
  typeFn?: ReturnTypeFunc;
  // TODO: this is for CatalogProductPropertyWithValue so we can handle it like the other fields
  ormTypeFn?: ReturnTypeFunc;
  name?: string;
  /**
   * Schema instructions.
   *  If set to `true`, the field will be hidden from the create operation, so the field cannot be set during creation.
   */
  hideCreate?: boolean;
  /**
   * If set to `true`, the field will be hidden from the update operation, so the field cannot be updated.
   */
  hideUpdate?: boolean;
  /**
   * Schema instructions.
   * If set to `true`, the field will be hidden from the where input type, so the field cannot be filtered.
   */
  hideWhere?: boolean;
  /**
   * Schema instructions.
   * If set to `true`, the relation can be connected by primary key
   */
  fieldOptions?: TypeOptions;
  /**
   * Relation schema instructions
   */
  relation?: CrudRelationFieldOptions & {
    parentPropert?: string;
  };
  staticWhereResolver?: WhereFieldResolver<Entity, GqlWhere>;
  whereResolver?: WhereFieldResolver<Entity, GqlWhere>;
  inputResolver?: true | InputFieldResolver<Entity, GqlInput>;
  staticInputResolver?: InputFieldResolver<Entity, GqlInput>;
};
export type CrudFieldOptions<
  Entity = any,
  GqlInput = any,
  GqlWhere = any,
> = CrudFieldProps<Entity, GqlInput, GqlWhere> & { propertyKey: string };

export function CrudField<Entity = any, GqlInput = any, GqlWhere = any>(
  options: CrudFieldProps<Entity, GqlInput, GqlWhere> = {},
) {
  return function (target: any, propertyKey: string) {
    const meta: CrudFieldOptions<Entity, GqlInput, GqlWhere> = {
      name: propertyKey,
      ...options,
      propertyKey,
    };

    addToCache(target, propertyKey, meta);
    // Reflect.defineMetadata(
    //   CRUD_FIELD_METADATA_NAME,
    //   { name: propertyKey, ...options, propertyKey },
    //   target,
    //   propertyKey
    // );
  };
}

export function getCrudFields<DTO>(target: Type<DTO>): CrudFieldOptions[] {
  const options = getPrototypeChain(target).reduce((fields, Cls) => {
    const cached = cache.get(Cls.prototype);

    if (!cached) {
      return fields;
    }
    return uniqBy(
      Object.values(cached).reduce((allFields, field) => {
        return [...allFields, field];
      }, fields),
      'name',
    );
  }, [] as any);

  return options;
}
