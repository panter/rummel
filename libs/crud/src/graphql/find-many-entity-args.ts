import { ArgsType, Field, InputType, Int } from '@nestjs/graphql';
import { CrudEntityType } from './crud-types';
import { findSortInput } from './find-sort-input';
import { whereInput } from './find-where-input';

export const findManyEntityArgs = <T, NA extends string>(
  classRef: CrudEntityType<T, NA>,
) => {
  const name = classRef.name + 'FindManyArgs';
  const WhereType = whereInput(classRef);
  const SortByType = findSortInput(classRef);

  @ArgsType()
  @InputType(name, { isAbstract: false })
  class FindManyEntityArgsType {
    @Field(() => WhereType, { nullable: true })
    where: any;

    @Field(() => [SortByType], { nullable: true })
    orderBy: any;

    @Field(() => Int, { nullable: true })
    skip?: number;

    @Field(() => Int, { nullable: true })
    take?: number;
  }

  return FindManyEntityArgsType as any;
};
