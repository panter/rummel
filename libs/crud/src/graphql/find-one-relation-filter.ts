import { Field, InputType } from '@nestjs/graphql';
import { CrudEntityType } from './crud-types';
import { whereInput } from './find-where-input';
import { typesCache } from './types-cache';
import { getCrudInfosForType, getTypeName } from './utils';

export const findOneRelationFilter = <T, NA extends string>(
  classRef: CrudEntityType<T, NA>,
) => {
  const typeName = getTypeName(classRef);
  const name = typeName + 'OneRelationFilter';
  if (typesCache[name]) {
    return typesCache[name];
  }

  @InputType(name, { isAbstract: false })
  class RelationFilterArgsType {}
  typesCache[name] = RelationFilterArgsType;

  getCrudInfosForType(whereInput(classRef)).forEach((p) => {
    Object.defineProperty(RelationFilterArgsType.prototype, p.schemaName, {
      writable: true,
      configurable: true,
    });
    Field(() => p.typeFn(), p.options)(
      RelationFilterArgsType.prototype,
      p.schemaName,
    );
  });
  return RelationFilterArgsType;
};
