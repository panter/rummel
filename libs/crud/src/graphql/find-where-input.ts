import { Type } from '@nestjs/common';
import { ArgsType, Field, InputType } from '@nestjs/graphql';
import { isArray } from 'lodash';
import { findOneRelationFilter } from './find-one-relation-filter';
import {
  BoolFilter,
  DateTimeFilter,
  IntFilter,
  StringFilter,
} from './gql-filter-types';
import { typesCache } from './types-cache';
import { CrudInfo, getCrudInfosForType, getTypeName } from './utils';

const getFilterDesignType = (p: CrudInfo) => {
  if (
    p.crudOptions?.hideWhere ||
    (!p.crudOptions?.whereResolver && p.isVirtual)
  ) {
    return;
  }

  const designType = p.typeFn();

  if (
    (designType as any).name === 'String' ||
    (designType as any).name === 'ID'
  ) {
    return StringFilter;
  } else if ((designType as any).name === 'Boolean') {
    return BoolFilter;
  } else if (
    (designType as any).name === 'Int' ||
    (designType as any).name === 'BigInt' ||
    (designType as any).name === 'Number' ||
    (designType as any).name === 'Float'
  ) {
    return IntFilter;
  } else if ((designType as any).name === 'Date') {
    return DateTimeFilter;
  } else if (p.options.isArray) {
    return whereInput(designType as Type);
  } else if (typeof designType !== 'function') {
    return StringFilter; // TODO workaround for enums
  } else {
    if (isArray(designType)) {
      return whereInput(designType as unknown as Type);
    }
    return findOneRelationFilter(designType as Type);
  }
};

export const whereInput = <T>(classRef: Type<T>, pentityType?: Type<T>) => {
  const entityType = pentityType || classRef;
  try {
    const fields = getCrudInfosForType(entityType);
    const typeName = getTypeName(classRef);
    const name = typeName + 'WhereInput';

    if (typesCache[name]) {
      return typesCache[name];
    }

    @ArgsType()
    @InputType(name, { isAbstract: false })
    class WhereInput {
      @Field(() => [WhereInput], { nullable: true })
      AND?: WhereInput[] | null;

      @Field(() => [WhereInput], { nullable: true })
      OR?: WhereInput[] | null;

      @Field(() => [WhereInput], { nullable: true })
      NOT?: WhereInput[] | null;
    }

    typesCache[name] = WhereInput;

    fields.forEach((p) => {
      const filterDesignType = getFilterDesignType(p);

      if (filterDesignType) {
        Object.defineProperty(WhereInput.prototype, p.schemaName, {
          writable: true,
          configurable: true,
        });
        Field(() => filterDesignType, { nullable: true })(
          WhereInput.prototype,
          p.schemaName,
        );
      }
    });

    return WhereInput;
  } catch (error) {
    // TODO: error handling
  }
};
