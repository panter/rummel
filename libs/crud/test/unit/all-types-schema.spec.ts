import {
  GraphQLSchemaBuilderModule,
  GraphQLSchemaFactory,
  Resolver,
} from '@nestjs/graphql';
import { printSchema } from 'graphql/utilities';
import { Test, TestingModule } from '@nestjs/testing';
import {
  CreateOneResolver,
  DeleteOneResolver,
  FindManyResolver,
  FindOneResolver,
  ObjectRelationResolvers,
  UpdateOneResolver,
} from '../../src';
import { ALL_SUPPORTED_TYPES_GQL_SCHEMA } from './fixtures/all-supported-types.fixtures';
import { Dummy } from './fixtures/all-types-shema.fixtures';
import { Dummy2 } from './fixtures/all-types-schema2.fixtures';

@Resolver(() => Dummy)
export class CreateOneDummyResolver extends CreateOneResolver(Dummy) {}

@Resolver(() => Dummy)
export class UpdateOneDummyResolver extends UpdateOneResolver(Dummy) {}

@Resolver(() => Dummy)
export class FindOneDummyResolver extends FindOneResolver(Dummy) {}

@Resolver(() => Dummy)
export class FindManyDummyResolver extends FindManyResolver(Dummy) {}

@Resolver(() => Dummy)
export class DeleteOneDummyResolver extends DeleteOneResolver(Dummy) {}

describe('Graphql types', () => {
  let app: TestingModule;

  beforeAll(async () => {
    app = await Test.createTestingModule({
      imports: [GraphQLSchemaBuilderModule],
    }).compile();
  });

  it('should generate graphql schema with all fields', async () => {
    const gqlSchemaFactory = app.get(GraphQLSchemaFactory);
    const schema = await gqlSchemaFactory.create(
      [
        CreateOneDummyResolver,
        UpdateOneDummyResolver,
        DeleteOneDummyResolver,
        FindOneDummyResolver,
        FindManyDummyResolver,
        ...ObjectRelationResolvers(Dummy),
        ...ObjectRelationResolvers(Dummy2),
      ],
      {
        skipCheck: true, //to avoid QueryRoot type must be provided error
      },
    );
    expect(printSchema(schema)).toEqual(ALL_SUPPORTED_TYPES_GQL_SCHEMA);
  });
});
