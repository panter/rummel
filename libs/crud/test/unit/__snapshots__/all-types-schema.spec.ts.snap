// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Graphql types should generate graphql schema with all fields 1`] = `
"type Dummy {
  idProp: ID!
  intProp: Int!
  floatProps: Float!
  enumProp: DummyEnum!
  stringProp: String!
  optionalStringProp: String
  boolProp: Boolean!
  dateProp: DateTime!
  oneToManyProp(where: Dummy2WhereInput, orderBy: [Dummy2OrderByInput!], skip: Int, take: Int): [Dummy2!]!
  manyToManyProp(where: Dummy2WhereInput, orderBy: [Dummy2OrderByInput!], skip: Int, take: Int): [Dummy2!]!
}

enum DummyEnum {
  A
  B
}

"""
A date-time string at UTC, such as 2019-12-03T09:54:33Z, compliant with the date-time format.
"""
scalar DateTime

input Dummy2WhereInput {
  AND: [Dummy2WhereInput!]
  OR: [Dummy2WhereInput!]
  NOT: [Dummy2WhereInput!]
  idProp: StringFilter
  dummyProp: DummyOneRelationFilter
}

input StringFilter {
  equals: String
  in: [String!]
  nin: [String!]
  contains: String
  gt: String
  gte: String
  lt: String
  lte: String
  fulltext: String
  ne: String
  mode: SringQueryMode
}

enum SringQueryMode {
  default
  insensitive
}

input DummyOneRelationFilter {
  AND: DummyWhereInput
  OR: DummyWhereInput
  NOT: DummyWhereInput
  idProp: StringFilter
  intProp: IntFilter
  floatProps: IntFilter
  enumProp: DummyEnumEnumFilter
  stringProp: StringFilter
  optionalStringProp: StringFilter
  boolProp: BoolFilter
  dateProp: DateTimeFilter
}

input DummyWhereInput {
  AND: [DummyWhereInput!]
  OR: [DummyWhereInput!]
  NOT: [DummyWhereInput!]
  idProp: StringFilter
  intProp: IntFilter
  floatProps: IntFilter
  enumProp: DummyEnumEnumFilter
  stringProp: StringFilter
  optionalStringProp: StringFilter
  boolProp: BoolFilter
  dateProp: DateTimeFilter
  oneToManyProp: Dummy2WhereInput
  manyToManyProp: Dummy2WhereInput
}

input IntFilter {
  equals: Float
  gt: Float
  gte: Float
  lt: Float
  lte: Float
  ne: Float
}

input DummyEnumEnumFilter {
  equals: DummyEnum
  in: [DummyEnum!]
  nin: [DummyEnum!]
  contains: DummyEnum
  gt: DummyEnum
  gte: DummyEnum
  lt: DummyEnum
  lte: DummyEnum
  fulltext: DummyEnum
  ne: DummyEnum
  mode: SringQueryMode
}

input BoolFilter {
  equals: Boolean
}

input DateTimeFilter {
  equals: DateTime
  gt: DateTime
  gte: DateTime
  lt: DateTime
  lte: DateTime
  ne: DateTime
}

input Dummy2OrderByInput {
  idProp: SortOrder
  dummyProp: DummyOrderByInput
}

enum SortOrder {
  asc
  desc
}

input DummyOrderByInput {
  idProp: SortOrder
  intProp: SortOrder
  floatProps: SortOrder
  enumProp: SortOrder
  stringProp: SortOrder
  optionalStringProp: SortOrder
  boolProp: SortOrder
  dateProp: SortOrder
  oneToManyProp: Dummy2OrderByInput
  manyToManyProp: Dummy2OrderByInput
}

type Dummy2 {
  idProp: ID!
  dummyProp: Dummy!
}

type Query {
  dummy(where: EntityIdInput): Dummy
  dummies(where: DummyWhereInput, orderBy: [DummyOrderByInput!], skip: Int, take: Int): [Dummy!]!
  dummiesCount(where: DummyWhereInput, orderBy: [DummyOrderByInput!], skip: Int, take: Int): Int!
}

input EntityIdInput {
  id: String!
}

type Mutation {
  createOneDummy(data: DummyCreateInput): Dummy!
  updateOneDummy(data: DummyUpdateInput, where: EntityIdInput!): Dummy!
  deleteOneDummy(where: EntityIdInput): Dummy!
}

input DummyCreateInput {
  intProp: Float
  floatProps: Float
  enumProp: DummyEnum
  stringProp: String
  optionalStringProp: String
  boolProp: Boolean
  dateProp: DateTime
  oneToManyProp: Dummy2ConnectCreateUpdateNestedManyWithoutDummyInput
  manyToManyProp: Dummy2ConnectCreateUpdateNestedManyWithoutDummyInput
}

input Dummy2ConnectCreateUpdateNestedManyWithoutDummyInput {
  connect: [ConnectRelationInput!]
  disconnect: [EntityIdInput!]
  create: [Dummy2CreateWithoutDummyInput!]
  update: [Dummy2UpdateWithWhereUniqueWithoutDummyInput!]
}

input ConnectRelationInput {
  id: String!
}

input Dummy2CreateWithoutDummyInput

input Dummy2UpdateWithWhereUniqueWithoutDummyInput {
  where: EntityIdInput!
  data: Dummy2UpdateInput
}

input Dummy2UpdateInput {
  dummyProp: DummyConnectNestedOneWithoutDummy2Input
}

input DummyConnectNestedOneWithoutDummy2Input {
  connect: ConnectRelationInput
  disconnect: Boolean
}

input DummyUpdateInput {
  intProp: NumberInput
  floatProps: NumberInput
  enumProp: DummyEnumEnumInput
  stringProp: StringInput
  optionalStringProp: StringInput
  boolProp: BooleanInput
  dateProp: DateInput
  oneToManyProp: Dummy2ConnectCreateUpdateNestedManyWithoutDummyInput
  manyToManyProp: Dummy2ConnectCreateUpdateNestedManyWithoutDummyInput
}

input NumberInput {
  set: Float
}

input DummyEnumEnumInput {
  set: DummyEnum
}

input StringInput {
  set: String
}

input BooleanInput {
  set: Boolean
}

input DateInput {
  set: DateTime
}"
`;
